File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/aas/backend/connected/aas/submodelelement/property/ConnectedMapProperty.java
package org.eclipse.basyx.aas.backend.connected.aas.submodelelement.property;

import java.util.Collection;
import java.util.Map;

import org.eclipse.basyx.aas.api.exception.ServerException;
import org.eclipse.basyx.aas.api.exception.TypeMismatchException;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.property.IMapProperty;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.property.PropertyType;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.submodelelement.property.SingleProperty;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;

/**
 * Connects to a PropertySingleValued as specified by DAAS containing a map
 * 
 * @author schnicke
 *
 */
public class ConnectedMapProperty extends ConnectedProperty implements IMapProperty {

	public ConnectedMapProperty(VABElementProxy proxy) {
		super(PropertyType.Map, proxy);
	}

	@Override
	public Object getValue(String key) throws TypeMismatchException, ServerException {
		return getMap().get(key);
	}

	@Override
	public void put(String key, Object value) throws ServerException {
		try {
			// check whether the value already exists and call update or create accordingly
			if(getValue(key) != null)
				getProxy().setModelPropertyValue(SingleProperty.VALUE + "/" + key, value);
			else
				getProxy().createValue(SingleProperty.VALUE + "/" + key, value);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public void set(Map<String, Object> map) throws ServerException {
		try {
			getProxy().setModelPropertyValue(SingleProperty.VALUE, map);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public Collection<String> getKeys() throws TypeMismatchException, ServerException {
		return getMap().keySet();
	}

	@Override
	public Integer getEntryCount() throws TypeMismatchException, ServerException {
		return getMap().entrySet().size();
	}

	@Override
	public void remove(String key) throws ServerException, TypeMismatchException {
		try {
			getProxy().deleteValue(SingleProperty.VALUE + "/" + key);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	private Map<String, Object> getMap() {
		return retrieveObject();
	}
}

File: sdks/c++/basys.sdk.cc/src/aas/aas/backend/connected/submodelelement/property/ConnectedMapProperty.cpp
/*
 * ConnectedMapProperty.cpp
 *
 *      Author: wendel
 */

#include "ConnectedMapProperty.h"
#include "vab/core/util/VABPath.h"

namespace basyx {
namespace aas {
namespace backend {
namespace connected {

using namespace submodelelement::property;

ConnectedMapProperty::ConnectedMapProperty(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedProperty(PropertyType::Map, proxy)
{}

basyx::any ConnectedMapProperty::getValue(std::string key)
{
  return this->getMap().at(key);
}

void ConnectedMapProperty::put(std::string key, basyx::any value)
{
  basyx::vab::core::VABPath path(PropertyPaths::VALUE);
  path.append(key);
  this->setProxyValue(path, value);
}

void ConnectedMapProperty::set(basyx::objectMap_t map)
{
  this->setProxyValue(PropertyPaths::VALUE, map);
}

basyx::objectCollection_t ConnectedMapProperty::getKeys()
{
  auto map = this->getMap();
  basyx::objectCollection_t keys;
  keys.reserve(map.size());

  for ( auto entry : map ) {
    keys.push_back(entry.first);
  }

  return keys;
}

int ConnectedMapProperty::getEntryCount()
{
  return this->getMap().size();
}

void ConnectedMapProperty::remove(std::string key)
{
  basyx::vab::core::VABPath path(PropertyPaths::VALUE);
  path.append(key);
  this->getProxy()->deleteElement(path);
}

basyx::objectMap_t ConnectedMapProperty::getMap()
{
  auto map = this->getProxy()->readElementValue(PropertyPaths::VALUE).Get<basyx::objectMap_t>();
  return map;
}

}
}
}
}
